Allow apps to customize NuttX build time variables

casaroli opened the pull request:
## Summary

This is the simplest way I could find to let the applications customize build-time options of NuttX (in `Make.defs`). For example, the `libtest` example is adding some libs to `EXTRA_LIBS`, and this is not being set.

I tried different approaches to do this and it looks like this is the less invasive of doing it in the current build system that I could come up with.

CMake is not impacted as it already correctly handles this situation.

## Impact

- Fixes https://github.com/apache/nuttx-apps/issues/2189
- `libtest` example apps can now tweak nuttx build-time variables
- Test that `libtest` builds in the CI

## Testing

- Enabled `libtest` example (`CONFIG_EXAMPLES_LIBTEST`) by default in `sim:module` config, so the CI will make sure this will keep working.
- Enable `libtest` example and the executable (`EXAMPLES_LIBTEST` and `EXAMPLES_LIBTEST_CMDTOOL`) and it works out of the box. I have tested for `esp32s3-devkit:nsh` and `sim:nsh`

acassis commented:
@casaroli @xiaoxiang781216 does it need to be extended to CMake as well?

acassis commented:
> > @casaroli @xiaoxiang781216 does it need to be extended to CMake as well?
> 
> I have never done a successful cmake build of NuttX.

I tested here and CMake is working fine to compile sim:nsh and others real boards

casaroli commented:
It looks like cmake works and do not need this modification.

However, in `libtest` we are missing the word `PRIVATE` in `target_sources`. I can do it in a separate PR

casaroli commented:
@xiaoxiang781216 do you think we can merge something like this?

xiaoxiang781216 commented on arch/arm/src/Makefile:
```diff
@@ -18,7 +18,7 @@
 ############################################################################
 
-include $(TOPDIR)/Make.defs
+include $(TOPDIR)/apps-or-nuttx-Make.defs
```
let's simple add:
```
-include $(APPDIR)/Make.defs
```
with the comment?

casaroli replied:
So the apps dir will be required even for a kernel only build of NuttX?

casaroli replied:
but then it won't have included `nuttx/Make.defs`.

The logic is "if there is `apps/Make.defs`, include it (which in turn includes `nuttx/Make.defs`, otherwise include `nuttx/Make.defs` directly"

xiaoxiang781216 replied:
Yes, you right

casaroli replied:
> Yes, you right. But can we do the trick inside Config.mk?

I tried, however the arch Makefiles include directly `nuttx/Make.defs` -- the board `Make.defs`, (which will eventually include `Config.mk`) and then `apps/Make.defs` would be included after we have already included `nuttx/Make.defs` so it won't work :(

Maybe a slightly cleaner solution is link the board `Make.defs` something else, like `nuttx/Board.defs` and then have this new file `apps-or-nuttx-Make.defs` be called `nuttx/Make.defs`, but this is essentially the same solution with better file naming.

casaroli commented:
I found another shorter solution to the problem and I will push it in a separate draft PR for your evaluation @acassis @xiaoxiang781216

xiaoxiang781216 commented:
> I found another shorter solution to the problem and I will push it in a separate draft PR for your evaluation @acassis @xiaoxiang781216

@casaroli do you have any update?

casaroli commented:
> > I found another shorter solution to the problem and I will push it in a separate draft PR for your evaluation @acassis @xiaoxiang781216
> 
> @casaroli do you have any update?

Hello, sorry for the delay.

Unfortunately I could not find a simpler portable solution to this, so the current proposal is the best I could come up with.

Maybe we should just rename the file `apps-or-nuttx-Make.defs` and put it some other place?

WDYT?

xiaoxiang781216 commented:
@casaroli but why do not include both $(TOPDIR)/Make.defs and $(APPDIR)/Make.defs?

casaroli commented:
> @casaroli but why do not include both (TOPDIR)/Make.defsand(APPDIR)/Make.defs?

I can't remember why it didn't work. I think some make variables that we are using `+=` got duplicated data. I will make another draft PR next week to try it.

casaroli commented:
> @casaroli but why do not include both (TOPDIR)/Make.defsand(APPDIR)/Make.defs?

You can see https://github.com/apache/nuttx/actions/runs/7897485926/job/21553195965?pr=11690 that the build failed because variables with repeated values.

casaroli commented:
I will update this just changing the same of the file and lets see if we can move on with this.

Otherwise I can just switch to cmake. But It would be nice if we can make `make` work.

casaroli commented:
I have moved the file to `tools` to reduce noise in the root dir.

I also added the `CONFIG_EXAMPLES_LIBTEST` to `sim:module` so we are sure that at least the build works in the CI.

Please let me know if we should add `CONFIG_EXAMPLES_LIBTEST` to `*:module` to test with all archs or if we should just not test the build and remove [the commit](https://github.com/apache/nuttx/pull/11202/commits/6676f924b61b78e7c0fa25ed77d855d372de66eb).

Thank you for your support.

xiaoxiang781216 commented:
@acassis please check carefully whether the patch pass ci before you merge it, merging ci failing patch happen many time.
